lib/posix/pthread.c

```diff
@@ -346,17 +346,14 @@ int pthread_getschedparam(pthread_t pthread, int *policy,
  */
 int pthread_once(pthread_once_t *once, void (*init_func)(void))
 {
-	pthread_mutex_lock(&pthread_key_lock);
+	k_spinlock_key_t key = k_spin_lock(&pthread_key_lock);
 
-	if (*once != PTHREAD_ONCE_INIT) {
-		pthread_mutex_unlock(&pthread_key_lock);
-		return 0;
+	if (*once == PTHREAD_ONCE_INIT) {
+		init_func();
+		*once = 0;
 	}
 
-	init_func();
-	*once = 0;
-
-	pthread_mutex_unlock(&pthread_key_lock);
+	k_spin_unlock(&pthread_key_lock, key);
```

Comment:
> Using a mutex in `pthread_once()` is probably overkill since it is not a cancellation point, according to the spec.

`init_func`, which is called inside the spin lock, is a cancellation point.

Reply:
> The pthread_once() function is not a cancellation point. However, if init_routine is a cancellation point and is canceled, the effect on once_control shall be as if pthread_once() was never called.

`init_routine` may be a cancellation point (it's not that it *is* by any means), but remediation in that case is most definitely not handled now either.

We can change this to a `struct k_mutex` if that makes things easier. My main goal here is to avoid using POSIX synchronisation types in the POSIX implementation.

Reply:
Given that `init_func` can be just about anything (from a cancellation point to a function that busy waits or iterates over hundreds of thousands of something), using a spin lock here would be a very dangerous thing.

`k_mutex` does sound a lot more reasonable.